packages/node/src/lib/get-group-id.ts

```diff
@@ -0,0 +1,203 @@
+import type { ApiKey, GroupingObject, KeyValue } from '../index';
+import type { Operation } from 'oas';
+
+import get from 'lodash/get';
+
+/**
+ * Parses the basic auth object to return the user
+ */
+const parseBasicAuth = (key: KeyValue): string | undefined => {
+  if (typeof key === 'object' && 'user' in key) {
+    return key.user;
+  }
+
+  if (typeof key === 'string') {
+    return key;
+  }
+
+  return undefined;
+};
+
+/**
+ * Finds the relevant key and its path for a given apiKey
+ */
+const findRelevantKey = (keys: ApiKey[], apiKey: string) => {
+  // Recursive function to traverse the nested structure of ApiKey and find the path to a specific key
+  const findPath = (obj: ApiKey | string | unknown, key: string, parentPath?: string): string | undefined => {
+    if (typeof obj === 'object' && obj !== null) {
+      return Object.keys(obj).reduce((acc: string | undefined, k) => {
+        // Construct the current path by appending the current key to the parent path
+        const currentPath = parentPath ? `${parentPath}.${k}` : k;
+        // Continue searching down the tree
+        return acc || findPath((obj as ApiKey)[k], key, currentPath);
+      }, undefined);
+    }
+
+    // If the current object is equal to the key, return the parent path
+    if (obj === key) {
+      return parentPath;
+    }
+
+    return undefined;
+  };
+
+  // Reduce the array of ApiKeys to find the relevant key and its path
+  return keys.reduce(
+    (acc: { relevantKey?: ApiKey; path?: string }, item) => {
+      const path = findPath(item, apiKey);
+      // If the path is found, return the relevantKey and path
+      if (path !== undefined) {
+        return { relevantKey: item, path };
+      }
+
+      return acc;
+    },
+    // Initialize the accumulator with the first key and undefined path
+    { relevantKey: keys[0], path: undefined },
+  );
+};
+
+/**
+ * Builds a flattened array of security requirements for an operation.
+ */
+const buildFlattenedSecurityRequirements = (operation: Operation): string[] => {
+  if (!operation) {
+    return [];
+  }
+
+  const securityRequirements = operation.getSecurity();
+  return securityRequirements.reduce((acc: string[], curr) => {
```

Comment:
just want to make sure I'm reading this right, but securityRequirements is an array of objects and then we reduce that to an array of all the keys on that array of objects? Does it matter that we may have duplicate entries of the same key?

Reply:
@jamestclark added this but I'm struggling to find how it is used. It's possible this existed in the main codebase and we reused parts of it in the sdk? However I don't think we have the OAS file at any point so I'm not sure

Reply:
Doesn't matter that we may have duplicates as we just loop through the array and find the first matching from the users apiKey.

And yeah it's not used directly by the metrics SDK. As the code is very similar to what we already had in the main codebase he idea here is to co-locate all the securityRequirements code here and have the main codebase use this once it's published.